### ISSUE TITLE
Controller fails VirtualService validation wth Istio webhook because of duplicate hosts in the spec

### ISSUE THREAD

**Issue by akirillov** [kubeflow/kfserving#948]
/kind bug

**What steps did you take and what happened:**
After the introduction of https://github.com/kubeflow/kfserving/pull/717, `VirtualServiceBuilder` adds duplicated hosts to `VistualService` spec and it fails Istio validation. Newly added code truncates `-predictor-default` suffix in [virtualservice.go#L314](https://github.com/kubeflow/kfserving/blob/master/pkg/controller/inferenceservice/resources/istio/virtualservice.go#L314) and returns hostname equal to the one returned by `network.GetServiceHostname(isvc.Name, isvc.Namespace)` in [virtualservice.go#L277-L280
](https://github.com/kubeflow/kfserving/blob/master/pkg/controller/inferenceservice/resources/istio/virtualservice.go#L277-L280)

The initial `InferenceService` received by KFServing controller:
```
{
	"kind": "InferenceService",
	"apiVersion": "serving.kubeflow.org/v1alpha2",
	"metadata": {
		"name": "mnist",
		"namespace": "test",
		"selfLink": "/apis/serving.kubeflow.org/v1alpha2/namespaces/test/inferenceservices/mnist",
		"uid": "4d18f854-50e8-4d41-bc17-5a1812c48fe3",
		"resourceVersion": "30770",
		"generation": 1,
		"creationTimestamp": "2020-07-14T20:55:54Z"
	},
	"spec": {
		"default": {
			"predictor": {
				"tensorflow": {
					"storageUri": "s3://mnist/mnist",
					"runtimeVersion": "1.14.0",
					"resources": {
						"limits": {
							"cpu": "1",
							"memory": "2Gi"
						},
						"requests": {
							"cpu": "1",
							"memory": "2Gi"
						}
					}
				}
			}
		}
	},
	"status": {
		"conditions": [{
			"type": "DefaultPredictorReady",
			"status": "True",
			"lastTransitionTime": "2020-07-14T22:10:33Z"
		}, {
			"type": "Ready",
			"status": "False",
			"lastTransitionTime": "2020-07-14T20:55:54Z",
			"reason": "PredictorHostnameUnknown",
			"message": "Failed to reconcile predictor"
		}, {
			"type": "RoutesReady",
			"status": "False",
			"lastTransitionTime": "2020-07-14T20:55:54Z",
			"reason": "PredictorHostnameUnknown",
			"message": "Failed to reconcile predictor"
		}],
		"default": {
			"predictor": {
				"name": "mnist-predictor-default-8fppb",
				"host": "mnist-predictor-default.test.svc.cluster.local"
			}
		},
		"canary": {}
	}
}
```

`VirtualService` constructed by KFServing controller (see `spec.hosts`):
```
{
	"metadata": {
		"name": "mnist",
		"namespace": "test",
		"creationTimestamp": null,
	},
	"spec": {
		"hosts": ["mnist.test.svc.cluster.local", "mnist.test.svc.cluster.local"],
		"gateways": ["kubeflow-gateway.kubeflow", "knative-serving/cluster-local-gateway"],
		"http": [{
			"match": [{
				"uri": {
					"prefix": "/v1/models/mnist:predict"
				},
				"authority": {
					"regex": "^mnist\\.test(\\.svc(\\.cluster\\.local)?)?(?::\\d{1,5})?$"
				},
				"gateways": ["kubeflow-gateway.kubeflow"]
			}, {
				"uri": {
					"prefix": "/v1/models/mnist:predict"
				},
				"authority": {
					"regex": "^mnist\\.test(\\.svc(\\.cluster\\.local)?)?(?::\\d{1,5})?$"
				},
				"gateways": ["knative-serving/cluster-local-gateway"]
			}],
			"route": [{
				"destination": {
					"host": "cluster-local-gateway.istio-system.svc.cluster.local",
					"port": {
						"number": 80
					}
				},
				"weight": 100,
				"headers": {
					"request": {
						"set": {
							"Host": "mnist-predictor-default.test.svc.cluster.local"
						}
					}
				}
			}],
			"retries": {
				"attempts": 3,
				"perTryTimeout": "600s"
			}
		}]
	}
}
```

Error from KFServing controller logs:
`admission webhook \"pilot.validation.istio.io\" denied the request: configuration is invalid: duplicate hosts in virtual service: mnist.test.svc.cluster.local & mnist.test.svc.cluster.local`

Full log message:
```
{"level":"error","ts":1594770415.6291103,"logger":"controller-runtime.controller","msg":"Reconciler error","controller":"kfserving-controller","request":"test/mnist","error":"admission webhook \"pilot.validation.istio.io\" denied the request: configuration is invalid: duplicate hosts in virtual service: mnist.test.svc.cluster.local & mnist.test.svc.cluster.local","stacktrace":"github.com/kubeflow/kfserving/vendor/github.com/go-logr/zapr.(*zapLogger).Error\n\t/go/src/github.com/kubeflow/kfserving/vendor/github.com/go-logr/zapr/zapr.go:128\ngithub.com/kubeflow/kfserving/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/go/src/github.com/kubeflow/kfserving/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:218\ngithub.com/kubeflow/kfserving/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/go/src/github.com/kubeflow/kfserving/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:192\ngithub.com/kubeflow/kfserving/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker\n\t/go/src/github.com/kubeflow/kfserving/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:171\ngithub.com/kubeflow/kfserving/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1\n\t/go/src/github.com/kubeflow/kfserving/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:152\ngithub.com/kubeflow/kfserving/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil\n\t/go/src/github.com/kubeflow/kfserving/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:153\ngithub.com/kubeflow/kfserving/vendor/k8s.io/apimachinery/pkg/util/wait.Until\n\t/go/src/github.com/kubeflow/kfserving/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88"}
```

**What did you expect to happen:**
`InferenceService` is created and gets into `Ready`  status

**Anything else you would like to add:**
Tried to run both manually by submitting a minimal required `InferenceService` spec and via KFP component - the result is the same.

Link to Istio 1.1.7 Pilot validation code: https://github.com/istio/istio/blob/1.1.7/pilot/pkg/model/validation.go#L1647-L1657.
Same validation code is present in newer versions (e.g. 1.6.4): https://github.com/istio/istio/blob/1.6.4/pkg/config/validation/validation.go#L2083-L2096 

**Environment:**

- Istio Version: 1.4.6, 1.5.4, 1.6.4
- Knative Version: 0.14.0
- KFServing Version: 0.3.0
- Kubeflow version: 1.0
- Kfdef:
- Minikube version:
- Kubernetes version: (use `kubectl version`): v1.16.4
- OS (e.g. from `/etc/os-release`):

**Comment by yuzisun** [comment]
@akirillov I can not replicate this, it should create one external host and the other one with internal host. Do you have some special configuration for knative or istio ?

**Comment by akirillov** [comment]
thanks for the quick response, @yuzisun! 

For Istio I have a default configuration with only mTLS and Ingress Gateway SDS enabled, `global.proxyclusterDomain` is set to the default `cluster.local`. KNative is installed with the default configuration. 

My guess would be that `spec.hosts` should look like that:
```
"hosts": ["mnist.test", "mnist.test.svc.cluster.local"],
```

Can you please point me at the code where `status.default.predictor` is getting populated so I can dig deeper?

**Comment by akirillov** [comment]
Did some digging into `ksvc` lifecycle, here's the findings.

Looking at the **KNative Serving** code base first. New `ksvc` creation is handled by the reconciler in [pkg/reconciler/service/service.go](https://github.com/knative/serving/blob/v0.14.0/pkg/reconciler/service/service.go). It creates a `Service` and a route [here](https://github.com/knative/serving/blob/v0.14.0/pkg/reconciler/service/service.go#L104):
```go
func (c *Reconciler) ReconcileKind(ctx context.Context, service *v1.Service) pkgreconciler.Event
...
config, err := c.config(ctx, logger, service)
...
route, err := c.route(ctx, logger, service)
...
        ss.PropagateRouteStatus(&route.Status)
...
```

After the route is created, its status is used to update the service status when it comes to the address:
```go
func (ss *ServiceStatus) PropagateRouteStatus(rs *RouteStatus) {
  ss.RouteStatusFields = rs.RouteStatusFields

  ...
}
```

Route creation in [pkg/reconciler/route/route.go](https://github.com/knative/serving/blob/v0.14.0/pkg/reconciler/route/route.go#L132-L137) has the following URL address-related code:
```go
r.Status.Address = &duckv1.Addressable{
    URL: &apis.URL{
        Scheme: "http",
        Host:   resourcenames.K8sServiceFullname(r),
    },
}
```
Here, `K8sServiceFullname` under the hood is calling `network.GetServiceHostname()` (same as in KFServing [virtualservice.go#L279](https://github.com/kubeflow/kfserving/blob/master/pkg/controller/inferenceservice/resources/istio/virtualservice.go#L279)).


Now, moving to **KFServing** code base. After the `ksvc` is created, the `InferenceService` status is updated in [inferenceservice_status.go#L147-L149](https://github.com/kubeflow/kfserving/blob/master/pkg/apis/serving/v1alpha2/inferenceservice_status.go#L147-L149):

```go
func (ss *InferenceServiceStatus) propagateStatus(...) {
  ...
  if serviceStatus.URL != nil {
    statusSpec.Hostname = serviceStatus.URL.Host
  }
  ...
```

After that, the status is used to extract the hostname and trim the suffix in [virtualservice.go#L309-L315](https://github.com/kubeflow/kfserving/blob/master/pkg/controller/inferenceservice/resources/istio/virtualservice.go#L309-L315):
```go
func getServiceHostname(isvc *v1alpha2.InferenceService) (string, error){
   predictorStatus, reason := getPredictStatusConfigurationSpec(isvc.Status.Default)
   ...
}
```

The bottom line is that the same function (`network.GetServiceHostname`) is used for both KNative route creation and for [adding the second hostname in KFServing](https://github.com/kubeflow/kfserving/blob/master/pkg/controller/inferenceservice/resources/istio/virtualservice.go#L279). The initial value from KNative Service Status is `mnist-predictor-default.namespace.svc.cluster.local` and the hostname created from `isvc` in KFServing is `mnist.namespace.svc.cluster.local`. After removing suffix from the `ksvc` hostname these values become equal and fail Istio validation.

**Comment by akirillov** [comment]
one major difference from the default config I have is the default `svc.cluster.local` domain set for KNative Serving but based on the code it doesn't seem to affect the behavior.
```
$> kubectl get cm config-domain --namespace knative-serving -o yaml

apiVersion: v1
data:
  svc.cluster.local: ""
kind: ConfigMap
metadata:
  labels:
    serving.knative.dev/release: v0.14.0
  name: config-domain
  namespace: knative-serving
```

**Comment by yuzisun** [comment]
@akirillov I see, that's why it caused the conflicts. KFServing assumes that is the external domain and we again add the rule for internal host name, so there needs to be a check.

**Comment by akirillov** [comment]
@yuzisun, shall we add a host duplicate check in `virtualservice.go`?

Also, can you please elaborate more on why the solution to the original problem was the changing of the hostname in case of `mnist-predictor-default`? The original issue had to deal with namespace erasure from the full hostname, but the current approach just _changes_ the hostname to `mnist` - even though it works for wildcard matching I'm not sure what are the consequences of the renaming.

**Comment by yuzisun** [comment]
@akirillov I think the issue #717 is irrelevant here, we use hostname name `mnist` since day 1 and  #717 is really addressing the a different domain template issue where we do not need to depend on the strong assumption that the domain is always in a format of `servicename.namespace` by deriving the url from the default mandatory predictor component. We use virtual host `mnist` because inferenceservice is not just one component, there are transformer and explainer, for one component you can have default and canary endpoint, the virtual host `mnist` is to handle the routing between these components and in addition it does traffic split between the default/canary endpoint. 

The conflicts issue I think is really caused by [here](https://github.com/kubeflow/kfserving/blob/master/pkg/controller/inferenceservice/resources/istio/virtualservice.go#L278), the hosts are duplicated, we expect the external host domain is different from the internal one but you are configuring your external domain the same as the internal in cluster domain. We can add a check but I am also curious the reason why you configure the external domain as internal one.

**Comment by akirillov** [comment]
Thanks for the detailed explanation, @yuzisun! So answering your question, we're defaulting all the inference services to be cluster-local and following the [KNative Serving documentation](https://knative.dev/docs/serving/cluster-local-route/):
>To make all services only cluster-local, change the default domain to svc.cluster.local by editing the config-domain config map. This will change all services deployed through Knative to only be published to the cluster, none will be available off-cluster.

Please correct me, if this is not how it should be done and if there's another way of achieving it.

**Comment by yuzisun** [comment]
@akirillov Got it,  it is the global config for `serving.knative.dev/visibility=cluster-local`, I agree this is a valid bug we should fix then.
